typing/types.mli

```diff
@@ -24,6 +24,105 @@
 (** Asttypes exposes basic definitions shared both by Parsetree and Types. *)
 open Asttypes
 
+module Sort : sig
+  (** A sort classifies how a type is represented at runtime. Every concrete
+      layout has a sort, and knowing the sort is sufficient for knowing the
+      calling convention of values of a given type. *)
+  type t
+
+  (** These are the constant sorts -- fully determined and without variables *)
+  type const =
+    | Void
+      (** No run time representation at all *)
+    | Value
+      (** Standard ocaml value representation *)
+
+  (** A sort variable that can be unified during type-checking. *)
+  type var
+
+  (** Create a new variable sort that can be unified. *)
+  val new_var : unit -> t
+
+  val of_const : const -> t
+  val of_var : var -> t
+
+  val void : t
+  val value : t
+
+  (** This checks for equality, and sets any variables to make two sorts
+      equal, if possible *)
+  val equate : t -> t -> bool
+end
+
+type sort = Sort.t
+
+module Layout : sig
+  (** A Layout.t is a full description of the runtime representation of values
+      of a given type. It includes sorts, but also the abstract top layout
+      [Any] and sublayouts of other sorts, such as [Immediate]. *)
+  type t
+
+  (** Constant layouts are used both for user-written annotations and within
+      the type checker when we know a layout has no variables *)
+  type const = Builtin_attributes.const_layout =
+    | Any
+    | Value
+    | Void
+    | Immediate64
+    | Immediate
+
+  val string_of_const : const -> string
+
+  (** This layout is the top of the layout lattice. All types have layout [any].
+      But we cannot compile run-time manipulations of values of types with layout
+      [any]. *)
+  val any : t
+
+  (** Value of types of this layout are not retained at all at runtime *)
+  val void : t
+
+  (** This is the layout of normal ocaml values *)
+  val value : t
+
+  (** Values of types of this layout are immediate on 64-bit platforms; on other
+      platforms, we know nothing other than that it's a value. *)
+  val immediate64 : t
+
+  (** We know for sure that values of types of this layout are always immediate *)
+  val immediate : t
+
+  (** Create a fresh sort variable, packed into a layout. *)
+  val new_var : unit -> t
```

Comment:
While the comment is clear, I'd prefer a name like `new_sort_var` or `any_sort`, mainly because it would make what's happening at use-sites clearer.  That is, unifying with a fresh type variable at layout `Layout.new_var`, when it occurs in `Typecore` or wherever, looks like it should always work (or at least not fail for layout reasons).  On the other hand, unifying with `Layout.new_sort_var ()` is more clearly checking that you have a representable layout.

Reply:
Yes, good idea

Reply:
I went with `of_new_sort_var`, because I like the ring of `Layout.of_new_sort_var`. It's a bit long, but I like it on balance.